pkg/apis/eventing/v1/broker_validation.go

```diff
@@ -33,36 +33,42 @@ const (
 
 func (b *Broker) Validate(ctx context.Context) *apis.FieldError {
 	ctx = apis.WithinParent(ctx, b.ObjectMeta)
-
 	cfg := config.FromContextOrDefaults(ctx)
-	var brConfig *config.ClassAndBrokerConfig
-	if cfg.Defaults != nil {
-		if c, ok := cfg.Defaults.NamespaceDefaultsConfig[b.GetNamespace()]; ok {
-			brConfig = c
-		} else {
-			brConfig = cfg.Defaults.ClusterDefault
-		}
-	}
 
-	withNS := ctx
-	if brConfig == nil || brConfig.DisallowDifferentNamespaceConfig == nil || !*brConfig.DisallowDifferentNamespaceConfig {
-		withNS = apis.AllowDifferentNamespace(ctx)
-	}
+	var errs *apis.FieldError
+	withNS := determineNamespaceAllowance(ctx, cfg.Defaults)
 
 	// Make sure a BrokerClassAnnotation exists
-	var errs *apis.FieldError
 	if bc, ok := b.GetAnnotations()[BrokerClassAnnotationKey]; !ok || bc == "" {
 		errs = errs.Also(apis.ErrMissingField(BrokerClassAnnotationKey))
 	}
 
+	// Further validation logic
 	errs = errs.Also(b.Spec.Validate(withNS).ViaField("spec"))
 	if apis.IsInUpdate(ctx) {
 		original := apis.GetBaseline(ctx).(*Broker)
 		errs = errs.Also(b.CheckImmutableFields(ctx, original))
 	}
+
 	return errs
 }
 
+// Determine if the namespace allowance based on the given configuration
+func determineNamespaceAllowance(ctx context.Context, brConfig *config.Defaults) context.Context {
+	if brConfig == nil || (brConfig.NamespaceDefaultsConfig == nil && brConfig.ClusterDefaultConfig == nil) {
+		return apis.AllowDifferentNamespace(ctx)
+	}
+	namespace := apis.ParentMeta(ctx).Namespace
+	nsConfig := brConfig.NamespaceDefaultsConfig[namespace]
+	if nsConfig == nil || nsConfig.DisallowDifferentNamespaceConfig == nil || !*nsConfig.DisallowDifferentNamespaceConfig {
+		if brConfig.ClusterDefaultConfig == nil || brConfig.ClusterDefaultConfig.DisallowDifferentNamespaceConfig == nil || !*brConfig.ClusterDefaultConfig.DisallowDifferentNamespaceConfig {
```

Comment:
Could you add a comment here explaining what's going on? Since the logic is quite complicated, it's hard to follow if you're not super familiar with the requirements of your new design (I could see myself messing it up if I had to modify this later)